### PR TITLE
stylo: Basic infrastructure for RestyleHint-driven traversal

### PR DESCRIPTION
**Pull request by bholley**
Gecko Bug: https://bugzilla.mozilla.org/show_bug.cgi?id=131701

(Don't review yet, will flag on the gecko bug when the time comes)

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14300)

### CODE REVIEW

**Comment by bholley** [comment]
@bors-servo retry

#13776

File: components/style/traversal.rs [hunk 1]
```diff
@@ -235,65 +225,38 @@ pub fn relations_are_shareable(relations: &StyleRelations) -> bool {
...
-{
-    let mut display_none = false;
-    ensure_element_styled_internal(element, context, &mut display_none);
-}
-
-#[allow(unsafe_code)]
-fn ensure_element_styled_internal<'a, E, C>(element: E,
-                                            context: &'a C,
-                                            parents_had_display_none: &mut bool)
+pub fn style_element_in_display_none_subtree<'a, E, C, F>(element: E,
```
Comment (emilio) [inline components/style/traversal.rs]: Would be neat to have a comment about his function, but I guess it's fine.

File: components/style/traversal.rs [hunk 2]
```diff
@@ -307,110 +270,236 @@ pub fn recalc_style_at<'a, E, C, D>(context: &'a C,
           D: DomTraversalContext<E::ConcreteNode>
 {
     // Get the style bloom filter.
+    //
```
Comment (emilio) [inline components/style/traversal.rs]: Gah, I see bitrots everywhere... :(

File: components/style/context.rs [hunk 3]
```diff
@@ -38,6 +38,10 @@ pub struct SharedStyleContext {
     /// Screen sized changed?
     pub screen_size_changed: bool,
 
+    /// Skip the root during traversal?
+    #[cfg(feature = "gecko")]
+    pub skip_root: bool,
```
Comment (emilio) [inline components/style/context.rs]: This needs a pretty big comment about why is it needed, if it's going to stay around.

Comment (emilio) [inline components/style/context.rs]: Well, I should stop reviewing without the gecko changes. Please add a note here than (as a followup) we can skip submitting the first job to the work queue (that would only contain the root) and send all the children instead.

File: components/layout/query.rs [hunk 4]
```diff
@@ -631,14 +618,24 @@ pub fn process_resolved_style_request<'a, N, C>(requested_node: N,
...
+
+    // We call process_resolved_style_request after performing a whole-document
+    // traversal, so the only reason we wouldn't have an up-to-date style here
+    // is that the requested node is in a display:none subtree. We currently
+    // maintain the invariant that elements in display:none subtrees always have
+    // no ElementData, so we need to temporarily bend those invariants here, and
+    // then throw them the style data away again before returning to preserve them.
+    // We could optimize this later to keep the style data cached somehow, but
+    // we'd need a mechanism to prevent detect when it's stale (since we don't
+    // traverse display:none subtrees during restyle).
```
Comment (emilio) [inline components/layout/query.rs]: Actually for stylo assuming we want to do that, I guess we can use either the undisplayed content map or the dom slots.

**Comment by emilio** [comment]
The changes look good. I guess I'd love a bunch more of documentation. I understand it _now_, after reading through the gecko patch thoroughly, and I think I don't have much more to say than what I said...

r=me on the servo-side if tests pass in servo.

File: components/style/traversal.rs [hunk 5]
```diff
@@ -62,7 +63,7 @@ pub fn take_thread_local_bloom_filter<E>(parent_element: Option<E>,
         match (parent_element, style_bloom.borrow_mut().take()) {
             // Root node. Needs new bloom filter.
             (None,     _  ) => {
-                debug!("[{}] No parent, but new bloom filter!", tid());
+                trace!("bloom: [{}] No parent, but new bloom filter!", tid());
```
Comment (bholley) [inline components/style/traversal.rs]: Done.

File: components/style/traversal.rs (hunk 2 above)
Comment (bholley) [inline components/style/traversal.rs]: Yeah, that's why I want to get this landed. :-(

**Comment by bholley** [comment]
Hm, I fixed that test and had it passing yesterday. Might be a mismerge, I'll look tomorrow noonish.

**Comment by emilio** [comment]
`./mach test-stylo` is failing because it's running without debug asertions and expects `Servo_AssertTreeIsClean` to be present :(

**Comment by bholley** [comment]
Doh!

@bors-servo r=emilio

**Comment by heycam** [comment]
@bors-servo retry
*  #13593

**Comment by KiChjang** [comment]
@bors-servo retry
* #13593

**Comment by KiChjang** [comment]
@bors-servo r-

The wrong test is disabled, the one that should be disabled is under `tests/wpt/metadata-css/css21-dev/html4/pseudo-elements-001.htm.ini`

**Comment by heycam** [comment]
@bors-servo retry force
* #13776

**Comment by bholley** [comment]
The last two failures were #14323 and #14317 respectively.

**Pull request merged**